---
title: Projecting scRNAseq data onto a reference map of Tumour-Infiltrating Lymphocytes
author: 
- Massimo Andreatta^[massimo.andreatta@unil.ch]
- Santiago Carmona^[santiago.carmona@unil.ch]
date: "09/04/2021"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
#output: html_notebook
editor_options: 
  chunk_output_type: inline
---

```{r message=F, warning=F, results=F, echo=F}
if (!requireNamespace("renv")) 
  install.packages("renv")
library(renv)
#renv::restore()
```

First, check package dependencies and install ProjecTILs

```{r message=F, warning=F}
Sys.setenv(R_REMOTES_NO_ERRORS_FROM_WARNINGS="true")

install.packages("desc")
library(desc)
library(Seurat)
#install.packages("remotes")
#library(remotes)

```


```{r}
#remotes::install_github("carmonalab/STACAS")
#remotes::install_github("carmonalab/ProjecTILs")
#install.packages("devtools")
library(devtools)
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ComplexHeatmap")
#devtools::install_version("crossmatch", version = "1.3.1", repos = "http://cran.us.r-project.org")
#devtools::install_version("multicross", version = "2.1.0", repos = "http://cran.us.r-project.org")
devtools::install_github("jackbibby1/SCPA")
library(SCPA)
```


#Load the Package and test the installation
```{r}
library(ProjecTILs)
ref <- load.reference.map()
data(query_example_seurat)

query.projected <- Run.ProjecTILs(query_example_seurat, ref=ref)

```

### Load reference atlas and query data

First, load the default reference TIL atlas.
If no reference map file is provided, the function `load.reference.map()` will automatically download it from  [https://doi.org/10.6084/m9.figshare.12478571](https://doi.org/10.6084/m9.figshare.12478571)


Let's explore the reference atlas
```{r}
refCols <- c("#edbe2a", "#A58AFF", "#53B400", "#F8766D", "#00B6EB", "#d1cfcc", "#FF0000", "#87f6a5", "#e812dd")
DimPlot(ref,label = T, cols = refCols)
```

See expression of important marker genes across reference subtypes
```{r}
markers <- c("Cd4","Cd8a","Ccr7","Tcf7","Pdcd1","Havcr2","Tox","Izumo1r","Cxcr6","Xcl1","Gzmb","Gzmk","Ifng","Foxp3")
VlnPlot(ref,features=markers,stack = T,flip = T,assay = "RNA")
```


Now let's load a query dataset - [Miller et al., Nature Immunol (2019)](https://pubmed.ncbi.nlm.nih.gov/30778252/)
```{r warning=FALSE}
#A sample data set is provided with the ProjecTILs package
querydata <- zsme3
#querydata <- ProjecTILs::query_example_seurat
```

More generally, it is possible to load a query matrix with gene names and barcodes (e.g. 10X format or raw counts)

##Raw count matrix from GEO
#BiocManager::install("GEOquery")

library(GEOquery)
geo_acc <- "GSE86028"
getGEOSuppFiles(geo_acc)

fname3 <- sprintf("%s/GSE86028_TILs_sc_wt_mtko.tpm.log2.txt.gz", geo_acc)
querydata3 <- read.sc.query(fname3, type = "raw.log2")


### Run Projection algorithm
```{r warning=FALSE,message=FALSE}
query.projected <- make.projection(querydata, ref=ref, filter.cells = FALSE, skip.normalize=TRUE)

```

NB: by default, `make.projection()` will pre-filter T cells using [scGate](https://github.com/carmonalab/scgate). In case the input dataset is already pre-filtered, or if you are using a non-T cell reference atlas, you can disable this step using `make.projection(querydata, ref=ref, filter.cells = FALSE)`.

Plot projection of new data over the reference in UMAP space. The contour lines display the density of projected query cells onto the reference map.
```{r warning=FALSE,message=FALSE}
DimPlot(querydata)
plot.projection(ref, query.projected)
```
##while maintaining my clusters


```{r}
plot.projection(ref, query.projected, pointsize = 0)

```
```{r}
query.classified

```

### Predict cell states

Predict the cell states in the query set using a nearest-neighbor algorithm
```{r warning=FALSE,message=FALSE}

query.projected <- cellstate.predict(ref=ref, query=query.projected)
table(query.projected$functional.cluster)

```

#Testing via Violin PLot

```{r}
VlnPlot(zsme.refined.2, features = "Dapl1", group.by = "functional.cluster", split.by = "orig.ident", assay = "RNA", add.noise = FALSE, pt.size = 0.1 )


```


```{r}
FeaturePlot(query.projected, features = "Dapl1", split.by = "orig.ident", pt.size = 2)
DimPlot(query.projected, group.by = "functional.cluster", split.by = "orig.ident", pt.size = 2)
DimPlot(query.classified, group.by = "functional.cluster", split.by = "orig.ident")
FeaturePlot(query.classified, features = "Dapl1", split.by = "orig.ident")

```
#refine our data based on the reference map

Let's remove the small contaminants we have from our CD4 cell subsets
```{r}
zsme.refined <- subset(query.projected, subset = functional.cluster == "Th1", invert = TRUE)
zsme.refined <- subset(zsme.refined, subset = functional.cluster == "Treg", invert = TRUE)
zsme.refined <- subset(zsme.refined, subset = functional.cluster == "CD4_NaiveLike", invert = TRUE)
zsme.refined <- subset(zsme.refined, subset = functional.cluster == "Tfh", invert = TRUE)
zsme.refined
```

#Replot
```{r}
#DimPlot(zsme.refined.2, group.by = "functional.cluster", split.by = "orig.ident")
DimPlot(zsme.refined, group.by = "functional.cluster", split.by = "orig.ident")
FeaturePlot(zsme.refined, features = "Dapl1", split.by = "orig.ident", pt.size = 1)

```

Plot the predicted composition of the query in terms of reference T cell subtypes
```{r warning=FALSE,message=FALSE}
#for ease of coding, I'm recharacterizing query.projected as the refined dataset 
query.projected <- zsme.refined
plot.statepred.composition(ref, query.projected,metric = "Percent")
```

How do the gene expression levels compare between reference and query for the different cell states?
```{r fig.height=8, fig.width=12, warning=FALSE,message=FALSE}

plot.states.radar(ref, query=query.projected, min.cells=30)
```

### Compare states across conditions

If we have multiple conditions (e.g. control vs. treatment, or samples from different tissues), we can search for discriminant genes between conditions (otherwise, by default this analysis is performed against the reference subtype as the 'control')



```{r fig.height=8, fig.width=12}
#Simulate a condition which e.g. increases Gzmb expression compared to control
query.control <- subset(query.projected, subset= orig.ident == "WT")
query.perturb <- subset(query.projected, subset= orig.ident == "Dapl1KO")

plot.states.radar(ref, query=list("Control" = query.control, "Query" = query.perturb))
```

In this toy example, where we simulated a condition that increases Gzmb expression compared to control, we expect cytotoxicity genes to drive differences.
```{r warning=FALSE,message=FALSE}

discriminantGenes <- find.discriminant.genes(ref=ref, query=query.perturb, query.control=query.control)

head(discriminantGenes,n=10)
```




We can use a volcano plot to display differentially expressed genes:
```{r}
library(EnhancedVolcano)
EnhancedVolcano(discriminantGenes, lab = rownames(discriminantGenes), x = "avg_log2FC", y = "p_val", pCutoff = 1e-09, 
    FCcutoff = 0.5, labSize = 5, legendPosition = "none", drawConnectors = F, title = "Gzmb_high vs. Gzmb_low (Tex)")
```


### Find discriminant dimensions

The dimensions in UMAP space summarize the main axes of variability of the reference map. What if the query data contains novel states? We can search for additional, maximally discriminant dimensions (either in ICA or PCA space) that explain new variability in the query set.

As before, simulate a condition which increases Gzmb expression compared to control

#for all D1 objects
```{r}
zsme3 <- query.projected

Dapl1KO.subset <- subset(x = zsme3, subset = orig.ident == "Dapl1KO")
WT.subset <- subset(x = zsme3, subset = orig.ident == "WT")
  
Dapl1_expression = GetAssayData(object = zsme3, 
assay = "RNA", slot = "data")["Dapl1",]

pos = names(which(Dapl1_expression>0))
neg = names(which(Dapl1_expression==0))

D1.pos_cells = subset(zsme3,cells=pos)
D1.neg_cells = subset(zsme3,cells=neg)

```


##Dapl1KO cluster

```{r}
Dapl1_expression.D1 = GetAssayData(object = Dapl1KO.subset, 
assay = "RNA", slot = "data")["Dapl1",]

pos_ids.D1 = names(which(Dapl1_expression.D1>0))
neg_ids.D1 = names(which(Dapl1_expression.D1==0))

Dapl1.pos_cells.D1 = subset(Dapl1KO.subset,cells=pos_ids.D1)
Dapl1.neg_cells.D1 = subset(Dapl1KO.subset,cells=neg_ids.D1)
```


```{r}
##WT T Cell Cluster
Dapl1_expression.WT = GetAssayData(object = WT.subset, 
assay = "RNA", slot = "data")["Dapl1",]

pos_ids.WT = names(which(Dapl1_expression.WT>0))
neg_ids.WT = names(which(Dapl1_expression.WT==0))

Dapl1.pos_cells.WT = subset(WT.subset,cells=pos_ids.WT)
Dapl1.neg_cells.WT = subset(WT.subset,cells=neg_ids.WT)

```



```{r}
table(Dapl1.pos_cells.D1$functional.cluster)
table(Dapl1.pos_cells.WT$functional.cluster)


```

##For Naive Like CD8's
```{r}
query.control.naive <- subset(Dapl1.pos_cells.WT, subset = functional.cluster == "CD8_NaiveLike")

query.perturb.naive <- subset(Dapl1.pos_cells.D1, subset = functional.cluster == "CD8_NaiveLike")

discriminantGenes.naive <- find.discriminant.genes(ref=ref, query=query.perturb, query.control=query.control)

discriminantGenes.naive

```

##For Tpex

```{r}
query.control.tpex <- subset(Dapl1.pos_cells.WT, subset = functional.cluster == "CD8_Tpex")

query.perturb.tpex <- subset(Dapl1.pos_cells.D1, subset = functional.cluster == "CD8_Tpex")

discriminantGenes.tpex <- find.discriminant.genes(ref=ref, query=query.perturb, query.control=query.control)

discriminantGenes.tpex


```


##Neat 3d modeling things

In this toy example, we expect some gene module associated with granzymes to drive the discriminant analysis:
```{r warning=FALSE,message=FALSE}
library(fastICA)

top.ica.wcontrol <- find.discriminant.dimensions(ref=ref, query=query.perturb, query.control=query.control)
```

```{r}
head(top.ica.wcontrol)
VizDimLoadings(ref, reduction = "ica", nfeatures = 10, dims=c(26,24,42), ncol=3)
```

Now we can plot the ICA dimension that captured the genetic changes associated to the perturbation of increasing Gzmb

```{r warning=FALSE,message=FALSE}
plot3d <- plot.discriminant.3d(ref, query=query.perturb, query.control=query.control, extra.dim="ICA_42")
plot3d

```


Focus the plot only on a specific state
r warning=FALSE,message=FALSE
plot3d <- plot.discriminant.3d(ref, query.projected, extra.dim="cycling.score", query.state="Tpex")
plot3d


##Redoing SPCA on subsets

```{r}
library(SCPA)
```

#WT Dapl1 positive cells
```{r}

D1.pos.wt.naive <- seurat_extract(
   D1.pos_cells,
   meta1 = "orig.ident",
   value_meta1 = "WT",
   meta2 = "functional.cluster",
   value_meta2 = "CD8_NaiveLike"
)

```

#Dapl1KO Dapl1 "wannabe" cells
```{r}

D1.pos.wb.naive <- seurat_extract(
   D1.pos_cells,
   meta1 = "orig.ident",
   value_meta1 = "Dapl1KO",
   meta2 = "functional.cluster",
   value_meta2 = "CD8_NaiveLike"
)
```

#Load Pathways
```{r}
pathways <- "G:\\.shortcut-targets-by-id\\1m88S3ZYWgmWlzsJb2wQneO8cakmw4oRT\\AL Lab computer\\AL-24\\mouse_H_v5p2_Lynn061417v2.csv"

```


#Compare Populations, for all DEG's in naive wb/wt cells


```{r}

scpa_out <- compare_pathways(samples = list(D1.pos.wb.naive, D1.pos.wt.naive),
                             pathways = pathways, 
                             downsample = 950)
scpa_out

```

# A heatmap, by "most significant pathways", and ignorign direction

```{r}

plot_heatmap(scpa_out, 
             show_row_names = T)

```

#Better visualizing the output


#for FDR values of 0, set threshold FDR to 0.001

```{r}
dataset <- scpa_out
dataset$qval[dataset$qval == 0] <- 0.001
#dataset$log2qval <- log2(dataset$qval)

```


```{r}
###using a divergent bar plot
ggplot(dataset, aes(x = reorder(Pathway, qval), y = FC,)) +
  geom_bar(stat = "identity",
           show.legend = FALSE,
           aes(fill = FC),  # Background color
           color = "gray30") + # Border color
  xlab("Group") +
  ylab("Value") +
  scale_fill_gradient2(low = "#6495ED",
                       mid = "aliceblue",
                       high = "#F4A460") +
  coord_flip() +
  theme_minimal()
  
```
###using a bubble chart


```{r}
ggplot(dataset, aes(x = reorder(Pathway, FC), y = FC, size = qval)) +
  geom_point(aes(fill = FC, color = FC), alpha = 1, stroke = 1) +
  scale_fill_gradient2(low = "#6495ED", mid = "aliceblue", high = "#F4A460") +
  scale_color_gradient2(low = "#6495ED", mid = "aliceblue", high = "#F4A460") +
  scale_size(range = c(min(dataset$qval), max(dataset$qval)), name = bquote(qval)) +
  xlab("Enrichment Term") + 
  ylab("Fold Change") +
  coord_flip()
```


```{r}
fc_range <- range(dataset$FC, na.rm = TRUE)
fc_breaks <- seq(from = fc_range[1], to = fc_range[2], length.out = 4)
fc_values <- rescale(fc_breaks, to = c(0,1))

ggplot(dataset, aes(x = reorder(Pathway, FC), y = FC, size = qval)) +
  geom_point(aes(fill = FC), alpha = 0.5, color = "black", shape = 21) +
  scale_fill_gradientn(colors = c("#6495ED", "aliceblue", "#F4A460"), values = fc_values) +
  scale_size(range = c(min(dataset$qval), max(dataset$qval)) * 1.5, name = bquote(qval)) +
  xlab("Enrichment Term") + 
  ylab("Fold Change") +
  coord_flip()


```
```{r}

library(dplyr)

filtered_data <- dataset %>% filter(qval > 1)

ggplot(filtered_data, aes(x = reorder(Pathway, FC), y = FC, size = qval)) +
  geom_point(aes(fill = FC), alpha = 0.5, color = "black", shape = 21) +
  scale_fill_gradientn(colors = c("#6495ED", "aliceblue", "#F4A460"), values = fc_values) +
  scale_size(range = c(min(filtered_data$qval), max(filtered_data$qval)) * 1.5, name = bquote(qval)) +
  xlab("Enrichment Term") + 
  ylab("Fold Change") +
  coord_flip()



```


##Trying again with Tpex cells

Compare Populations, for all DEG's in Tpex wb/wt cells



#WT Dapl1 positive cells
```{r}

D1.pos.wt.tpex <- seurat_extract(
   D1.pos_cells,
   meta1 = "orig.ident",
   value_meta1 = "WT",
   meta2 = "functional.cluster",
   value_meta2 = "CD8_Tpex"
)

```

#Dapl1KO Dapl1 "wannabe" cells
```{r}

D1.pos.wb.tpex <- seurat_extract(
   D1.pos_cells,
   meta1 = "orig.ident",
   value_meta1 = "Dapl1KO",
   meta2 = "functional.cluster",
   value_meta2 = "CD8_Tpex"
)
```


```{r}

scpa_out2 <- compare_pathways(samples = list(D1.pos.wb.tpex, D1.pos.wt.tpex),
                             pathways = pathways, 
                             downsample = 1000)
scpa_out2

```


#Doing seurat DEG things



#naming our subcats


```{r}

##WT positives vs. the "Dapl1KO" positive cells

Naive_wb_vs_control <- FindMarkers(
  zsme3,
  ident.1 = pos_ids.wb.naive,
  ident.2 = pos_ids.wt.naive,
  assay = "RNA",
  slot = "data",
  reduction = NULL,
  features = NULL,
  logfc.threshold = 0.25,
  test.use = "wilcox",
  min.pct = 0.1,
  min.diff.pct = 0.1,
  verbose = TRUE,
  only.pos = FALSE,
  max.cells.per.ident = 500,
  random.seed = 1,
  latent.vars = NULL,
  min.cells.feature = 3,
  min.cells.group = 3,
  mean.fxn = NULL,
  fc.name = NULL,
  base = 2,
  densify = FALSE,)

Naive_wb_vs_control

```


#highlighting interferon pathways


```{r}

# 4. highlight ifn pathways in naive cd4 -------------------------------------
pws <- c("INTERFE", "IFN")
ifn_paths <- hm_mat %>%
  rownames_to_column("pathway") %>%
  filter(grepl(paste(pws, collapse = "|"), pathway)) %>%
  .[1:6, ] %>%
  pull(pathway)

ifn_rank <- hm_mat %>%
  rownames_to_column("pathway") %>%
  arrange(desc(qval_n4)) %>%
  mutate(path_rank = percent_rank(qval_n4))

ggplot(ifn_rank, aes(qval_n4, path_rank*100)) +
  geom_hline(yintercept = c(0, 25, 50, 75, 100), linetype = 'dotted', lwd = 0.3, color = 'gray40') +
  geom_point(shape = 21, cex = 2, color = 'black', fill = 'royalblue2', stroke = 0.05) +
  geom_point(data = subset(ifn_rank, pathway %in% ifn_paths), shape = 21, cex = 3, color = 'black', fill = 'orangered2') +
  xlab("Qval") +
  ylab("Pathway rank") +
  ggtitle("Naive CD4 T cells") +
  scale_y_continuous(expand = c(0.03, 0.03), breaks = c(0, 25, 50, 75, 100)) +
  scale_x_continuous(expand = c(0.2, 0.2)) +
  theme(panel.border = element_rect(fill = NA),
        panel.background = element_blank(),
        title = element_text(size = 9),
        axis.title = element_text(size = 10))




